### PR TITLE
feat(modifier-managers): use 3.22 capabilities

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
BREAKING CHANGE: destruction behavior changed to use new args proxy from 3.22

Resolves the issue
  described by: https://github.com/emberjs/ember.js/issues/19162
  fixed by: https://github.com/emberjs/ember.js/pull/19163

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -49,9 +49,7 @@ jobs:
     # we recommend new addons test the current and previous LTS
     # as well as latest stable release (bonus points to beta/canary)
     - stage: "Additional Tests"
-      env: EMBER_TRY_SCENARIO=ember-lts-2.18
-    - env: EMBER_TRY_SCENARIO=ember-lts-3.4
-    - env: EMBER_TRY_SCENARIO=ember-release
```
Comment (rwjblue) [inline .travis.yml:54]: Need to bring these back, along with adding 3.8, 3.12, 3.16, and 3.20.

File: addon/modifiers/did-insert.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ import { setModifierManager, capabilities } from '@ember/modifier';
 */
 export default setModifierManager(
   () => ({
-    capabilities: capabilities('3.13', { disableAutoTracking: true }),
```
Comment (rwjblue) [inline addon/modifiers/did-insert.js:49]: `disableAutoTracking` was intentional (and is what the RFC specifies the behavior of these modifiers should be)

File: addon/modifiers/did-update.js [hunk 3]
```diff
@@ -58,7 +58,7 @@ import { setModifierManager, capabilities } from '@ember/modifier';
 */
 export default setModifierManager(
   () => ({
-    capabilities: capabilities('3.13', { disableAutoTracking: true }),
```
Comment (rwjblue) [inline addon/modifiers/did-update.js:61]: `disableAutoTracking` was intentional (and is what the RFC specifies the behavior of these modifiers should be)

File: addon/modifiers/will-destroy.js [hunk 4]
```diff
@@ -40,7 +40,7 @@ import { setModifierManager, capabilities } from '@ember/modifier';
 */
 export default setModifierManager(
   () => ({
-    capabilities: capabilities('3.13', { disableAutoTracking: true }),
```
Comment (rwjblue) [inline addon/modifiers/will-destroy.js:43]: `disableAutoTracking` was intentional (and is what the RFC specifies the behavior of these modifiers should be)

File: package.json [hunk 5]
```diff
@@ -22,8 +22,7 @@
     "test:all": "ember try:each"
   },
   "dependencies": {
-    "ember-cli-babel": "^7.10.0",
-    "ember-modifier-manager-polyfill": "^1.1.0"
```
Comment (rwjblue) [inline package.json:26]: Should come back when we bring back the other ember-source versions in ember-try scenarios

File: tests/integration/modifiers/did-update-test.js [hunk 6]
```diff
@@ -23,5 +23,6 @@ module('Integration | Modifier | did-update', function(hooks) {
     );
 
     this.set('boundValue', 'update');
+    await settled();
```
Comment (rwjblue) [inline tests/integration/modifiers/did-update-test.js:26]: Why is this needed?

**Review by rwjblue — CHANGES_REQUESTED** [review]
In general, this can be done without a breaking change. Bring in either ember-compatibility-helpers or `@embroider/macros` and use the appropriate manager API version based on the running Ember version.

Something like:

```js
import { gte } from 'ember-compatibility-helpers';
import { setModifierManager, capabilities } from '@ember/modifier';

export default setModifierManager(
  () => ({
    capabilities: capabilities(gte('3.22.0') ? '3.22' : '3.13', {
      disableAutoTracking: true,
    }),

    createModifier() {},

    installModifier(_state, element, args) {
      let [fn, ...positional] = args.positional;

      fn(element, positional, args.named);
    },

    updateModifier() {},
    destroyModifier() {},
  }),
  class DidInsertModifier {}
);
```

File: tests/integration/modifiers/did-update-test.js (hunk 6 above)
Comment (NullVoxPopuli) [reply tests/integration/modifiers/did-update-test.js:26]: it's not! I was stabbin' in the dark trying to change the test outcome

File: addon/modifiers/did-update.js [hunk 7]
```diff
@@ -58,7 +59,7 @@ import { setModifierManager, capabilities } from '@ember/modifier';
...
+    capabilities: capabilities(gte('3.22.0') ? '3.22' : '3.13', { disableAutoTracking: true }),
```
Comment (rwjblue) [inline addon/modifiers/did-update.js:62]: Need to chat with @pzuraq about this one (did-insert and will-destroy seem fine). Specifically RE: `disableAutoTracking`. If we move to using 3.22 version of modifier managers we would not get rerendered when our own args change (kinda the point of `did-update` IMHO).

I don't know what the intersection of using the args proxy **and** using `disableAutoTracking` will do...

Comment (rwjblue) [reply addon/modifiers/did-update.js:62]: My guess is that this is why the tests are failing in 3.22:

```
not ok 5 Chrome 62.0 - [61 ms] - Integration | Modifier | did-update: it basically works
    ---
        actual: >
            null
        stack: >
                at Object.<anonymous> (http://localhost:7357/assets/tests.js:167:21)
                at processModule (http://localhost:7357/assets/test-support.js:3963:17)
                at module$1 (http://localhost:7357/assets/test-support.js:3988:4)
                at Module.callback (http://localhost:7357/assets/tests.js:165:21)
                at Module.exports (http://localhost:7357/assets/vendor.js:111:32)
                at requireModule (http://localhost:7357/assets/vendor.js:32:18)
                at TestLoader.<anonymous> (http://localhost:7357/assets/test-support.js:14870:11)
                at TestLoader.require (http://localhost:7357/assets/test-support.js:14860:27)
        message: >
            Expected 4 assertions, but 0 were run
        negative: >
            false
        browser log: |
```

File: addon/modifiers/did-update.js [hunk 8]
```diff
@@ -58,14 +59,30 @@ import { setModifierManager, capabilities } from '@ember/modifier';
...
 
     createModifier() {
       return { element: null };
     },
-    installModifier(state, element) {
+
+    installModifier(instance, element, args) {
       // save element into state bucket
-      state.element = element;
+      instance.element = element;
```
Comment (rwjblue) [inline addon/modifiers/did-update.js:72]: Please revert this. The name of the first variable should not be `instance` (`bucket` is what we commonly use, but `state` is still better than `bucket`).

File: addon/modifiers/did-update.js (hunk 7 above)
Comment (NullVoxPopuli) [reply addon/modifiers/did-update.js]: I think I figured it out?

File: addon/modifiers/did-update.js (hunk 8 above)
Comment (NullVoxPopuli) [reply addon/modifiers/did-update.js]: done!

Comment (NullVoxPopuli) [reply addon/modifiers/did-update.js]: does this mean that this should change to state? https://github.com/ember-modifier/ember-modifier/blob/master/addon/-private/class/modifier-manager.ts#L32

**Comment by NullVoxPopuli** [comment]
travis build for ember 3.12 needs to be re-run I think? it crashed before the actual relevant-parts-to-this-PR began

**Comment by NullVoxPopuli** [comment]
green!

File: .travis.yml [hunk 9]
```diff
@@ -3,10 +3,9 @@ language: node_js
 node_js:
   # we recommend testing addons with the same minimum supported node version as Ember CLI
   # so that your addon works for all apps
-  - "6"
```
Comment (rwjblue) [inline .travis.yml:6]: Needs to land in the other PR first

File: addon/modifiers/did-update.js [hunk 10]
```diff
@@ -58,14 +59,30 @@ import { setModifierManager, capabilities } from '@ember/modifier';
...
+        // consume all args
+        for (let i = 0; i < args.positional.length; i++) {
+          // "noop" / consume the arg
+          args.positional[i];
+        }
+
+        for (let key of Object.keys(args.named)) {
+          // "noop" / consume the arg
+          args.named[key];
+        }
```
Comment (rwjblue) [inline addon/modifiers/did-update.js:84]: Ya, we definitely shouldn't do this. We should either stick with 3.13 capabilities (which does this for us) or fix the tests to actual consume. 

@pzuraq what did the RFC say about rerunning RE: the arguments changes? Does it mention auto tracking (and therefore infer that consumption is required)?

**Review by rwjblue — CHANGES_REQUESTED** [review]

File: addon/modifiers/did-update.js (hunk 10 above)
Comment (NullVoxPopuli) [reply addon/modifiers/did-update.js:84]: > Ya, we definitely shouldn't do this.

I don't think a did-update hook is possible to exist then :-/  (with capabilities 3.22)
it order for the `update` to run, tracked things _must_ be consumed during install.

Comment (NullVoxPopuli) [reply addon/modifiers/did-update.js:84]: this _very much a wip_, but I'm gonna be doing these sorts of tests for ember-modifier, too: https://github.com/NullVoxPopuli/ember-render-modifiers/pull/1

Comment (NullVoxPopuli) [reply addon/modifiers/did-update.js:84]: Where do you do that consumption? Did-update will not consume without running, and it won't run without an update. That leaves the only option to consume during install.

> Moved to draft state

I meant the linked PR :/

**Comment by NullVoxPopuli** [comment]
After talking with @pzuraq about this, I'm gonna close it :D 

It doesn't make sense to use 3.22 capabilities, since this addon is an anti-pattern in modern code anyway. :D

**Comment by SergeAstapov** [comment]
> this addon is an anti-pattern in modern code

@NullVoxPopuli do you mind to share some info about that statement? Seems waky to hear that about what guides do teach us in https://guides.emberjs.com/release/components/template-lifecycle-dom-and-modifiers/#toc_calling-methods-on-first-render
